src/ucp/core/ucp_rkey.c

```diff
@@ -179,16 +212,22 @@ UCS_PROFILE_FUNC(ssize_t, ucp_rkey_pack_memh,
                   md_index, context->tl_mds[md_index].rsc.md_name);
     }
 
-    if (ucs_likely(mem_info->sys_dev == UCS_SYS_DEVICE_ID_UNKNOWN)) {
-        goto out_packed_size;
+
+    if (legacy) {
+        if (ucs_likely(mem_info->sys_dev == UCS_SYS_DEVICE_ID_UNKNOWN)) {
+            goto out_packed_size;
+        }
+    } else {
+        /* Pack list of system device id */
+        *ucs_serialize_next(&p, ucp_sys_dev_map_t) = sys_dev_map;
     }
 
     /* Pack system device id */
     *ucs_serialize_next(&p, uint8_t) = mem_info->sys_dev;
 
     /* Pack distance from sys_dev to each device in distance_dev_map */
     ucs_for_each_bit(sys_dev, sys_dev_map) {
-        ucp_rkey_pack_distance(sys_dev, sys_distance++,
+        ucp_rkey_pack_distance(sys_dev, NULL, mem_info->sys_dev,
```

Comment:
Passing `NULL` here may cause an issue in case of rendezvous protocols, since `sys_dev_distance` is ignored.